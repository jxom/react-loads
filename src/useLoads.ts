import * as React from 'react';
import useDetectMounted from './hooks/useDetectMounted';
import useTimeout from './hooks/useTimeout';
import StateComponent from './StateComponent';
import { LoadsContext } from './LoadsContext';
import { LoadsConfig, LoadFunction, LoadingState, OptimisticCallback, OptimisticOpts, Record } from './types';

const STATES: { [key: string]: LoadingState } = {
  IDLE: 'idle',
  PENDING: 'pending',
  TIMEOUT: 'timeout',
  RESOLVED: 'resolved',
  REJECTED: 'rejected'
};

export default function useLoads<R>(
  fn: LoadFunction<R>,
  {
    cacheProvider,
    context: contextKey,
    delay = 300,
    enableBackgroundStates = false,
    unstable_enableSuspense,
    defer = false,
    loadPolicy = 'cache-and-load',
    timeout = 0,
    update: updateFn
  }: LoadsConfig<R> = {},
  inputs: Array<any> = []
) {
  const globalContext = React.useContext(LoadsContext);
  const counter = React.useRef<number>(0);
  const hasMounted = useDetectMounted();
  const loadsPromise = React.useRef<Promise<void> | void>(undefined);
  const [setDelayTimeout, clearDelayTimeout] = useTimeout();
  const [setTimeoutTimeout, clearTimeoutTimeout] = useTimeout();

  function reducer(state: Record<R>, action: { type: LoadingState; isCached?: boolean; response?: R; error?: any }) {
    switch (action.type) {
      case STATES.IDLE:
        return { state: STATES.IDLE };
      case STATES.PENDING:
        return { ...state, state: STATES.PENDING };
      case STATES.TIMEOUT:
        return { ...state, state: STATES.TIMEOUT };
      case STATES.RESOLVED:
        return { ...state, isCached: action.isCached, response: action.response, state: STATES.RESOLVED };
      case STATES.REJECTED:
        return { ...state, isCached: action.isCached, error: action.error, state: STATES.REJECTED };
      default:
        return state;
    }
  }

  const cachedRecord = React.useMemo(
    () => {
      if (contextKey) {
        return globalContext.cache.get(contextKey, { cacheProvider });
      }
      return;
    },
    [cacheProvider, contextKey, globalContext.cache]
  );

  let initialRecord = { state: STATES.IDLE };
  if (cachedRecord && !defer && loadPolicy !== 'load-only') {
    initialRecord = cachedRecord;
  }
  const [record, dispatch] = React.useReducer(reducer, initialRecord);

  function handleData(data: { response?: R; error?: any }, state: LoadingState, count: number) {
    if (hasMounted.current && count === counter.current) {
      // @ts-ignore
      clearDelayTimeout();
      // @ts-ignore
      clearTimeoutTimeout();
      dispatch({
        type: state,
        isCached: Boolean(contextKey),
        error: state === STATES.REJECTED ? data.error : undefined,
        response: state === STATES.RESOLVED ? data.response : undefined
      });
      if (contextKey) {
        const record = { error: data.error, response: data.response, state };
        globalContext.cache.set(contextKey, record, { cacheProvider });
      }
    }
  }

  function handleOptimisticData(
    {
      data,
      optsOrCallback,
      callback
    }: { data: any; optsOrCallback?: OptimisticOpts<R> | OptimisticCallback; callback?: OptimisticCallback },
    state: LoadingState,
    count: number
  ) {
    let newData = data;
    let opts: OptimisticOpts<R> = {};

    if (typeof optsOrCallback === 'object') {
      opts = optsOrCallback;
    }

    if (typeof data === 'function') {
      let cachedValue;
      if (record.response) {
        cachedValue = record.response;
      } else if (opts.context) {
        cachedValue = globalContext.cache.get(opts.context, { cacheProvider }) || {};
      }
      newData = data(cachedValue);
    }

    const value = {
      error: state === STATES.REJECTED ? newData : undefined,
      response: state === STATES.RESOLVED ? newData : undefined
    };
    if (!opts.context || contextKey === opts.context) {
      handleData(value, state, count);
    } else {
      if (globalContext.cache) {
        globalContext.cache.set(opts.context, { ...value, state }, { cacheProvider });
      }
    }

    let newCallback = typeof optsOrCallback === 'function' ? optsOrCallback : callback;
    newCallback && newCallback(newData);
  }

  function load(opts?: { fn?: LoadFunction<R> }) {
    return (..._args: any) => {
      const args = _args.filter((arg: any) => arg.constructor.name !== 'Class');

      counter.current = counter.current + 1;

      if (contextKey && loadPolicy !== 'load-only') {
        if (cachedRecord) {
          dispatch({ type: cachedRecord.state, isCached: true, ...cachedRecord });
          if (loadPolicy === 'cache-first') return;
        }
      }

      if (delay > 0) {
        setDelayTimeout(() => dispatch({ type: STATES.PENDING }), delay);
      } else {
        dispatch({ type: STATES.PENDING });
      }
      if (timeout > 0) {
        setTimeoutTimeout(() => dispatch({ type: STATES.TIMEOUT }), timeout);
      }

      const loadFn = opts && opts.fn ? opts.fn : fn;
      const promise = loadFn(...args, {
        cachedRecord,
        setResponse: (
          data: any,
          optsOrCallback: OptimisticOpts<R> | OptimisticCallback,
          callback?: OptimisticCallback
        ) => handleOptimisticData({ data, optsOrCallback, callback }, STATES.RESOLVED, counter.current),
        setError: (data: any, optsOrCallback: OptimisticOpts<R> | OptimisticCallback, callback?: OptimisticCallback) =>
          handleOptimisticData({ data, optsOrCallback, callback }, STATES.REJECTED, counter.current)
      })
        .then(response => handleData({ response }, STATES.RESOLVED, counter.current))
        .catch(err => handleData({ error: err }, STATES.REJECTED, counter.current));
      loadsPromise.current = promise;
      return promise;
    };
  }

  const update = React.useMemo(
    () => {
      if (!updateFn) return;
      if (Array.isArray(updateFn)) {
        return updateFn.map(fn => load({ fn }));
      }
      return load({ fn: updateFn });
    },
    [updateFn] // eslint-disable-line react-hooks/exhaustive-deps
  );

  React.useEffect(
    () => {
      if (cachedRecord && loadPolicy !== 'load-only') {
        dispatch({ type: cachedRecord.state, isCached: true, ...cachedRecord });
      }
    },
    [cachedRecord, loadPolicy, dispatch]
  );

  React.useEffect(
    () => {
      if (defer) return;
      load()();
    },
    [defer, contextKey, !inputs ? fn : undefined, ...inputs] // eslint-disable-line react-hooks/exhaustive-deps
  );

  const states = {
    isIdle: record.state === STATES.IDLE && Boolean((!record.response && !record.error) || enableBackgroundStates),
    isPending:
      record.state === STATES.PENDING && Boolean((!record.response && !record.error) || enableBackgroundStates),
    isTimeout:
      record.state === STATES.TIMEOUT && Boolean((!record.response && !record.error) || enableBackgroundStates),
    isResolved: record.state === STATES.RESOLVED || Boolean(record.response),
    isRejected: record.state === STATES.REJECTED || Boolean(record.error)
  };

  function isSuspenseEnabled() {
    if (unstable_enableSuspense === false) {
      return false;
    }
    if (unstable_enableSuspense) {
      return true;
    }
    if (globalContext.unstable_enableSuspense) {
      return true;
    }
    return false;
  }
  if (isSuspenseEnabled() && loadsPromise.current) {
    if (states.isPending) {
      throw loadsPromise.current;
    }
  }

  return React.useMemo(
    () => ({
      load: load(),
      update,

      response: record.response,
      error: record.error,
      state: record.state,

      ...states,
      Idle: StateComponent(states.isIdle),
      Pending: StateComponent(states.isPending),
      Timeout: StateComponent(states.isTimeout),
      Resolved: StateComponent(states.isResolved),
      Rejected: StateComponent(states.isRejected),

      isCached: Boolean(record.isCached)
    }),
    [record.response, record.error, record.state, record.isCached, states, update] // eslint-disable-line react-hooks/exhaustive-deps
  );
}
