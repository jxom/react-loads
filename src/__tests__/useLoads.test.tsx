import * as React from 'react';
import { render, waitForElement, wait, fireEvent } from 'react-testing-library';
import useLoads from '../useLoads';

describe('states', () => {
  it('renders idle correctly the loading function has not been invoked', () => {
    const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
    const Component = () => {
      const testLoader = useLoads(fn, { defer: true });
      return <React.Fragment>{testLoader.isIdle && 'idle'}</React.Fragment>;
    };

    const { container } = render(<Component />);

    expect(fn).toBeCalledTimes(0);
    expect(container).toMatchSnapshot();
  });

  describe('with load on initial render', () => {
    it('renders loading indicator correctly (default delay)', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn);
        return <React.Fragment>{testLoader.isPending && 'loading'}</React.Fragment>;
      };

      const { container, getByText } = render(<Component />);

      await waitForElement(() => getByText('loading'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders timeout indicator correctly', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn, { timeout: 400 });
        return (
          <React.Fragment>
            {testLoader.isPending && 'loading'}
            {testLoader.isTimeout && 'timeout'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('timeout'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders success correctly when the function has been resolved', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => res()));
      const Component = () => {
        const testLoader = useLoads(fn);
        return (
          <React.Fragment>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isResolved && 'success'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      await waitForElement(() => getByText('success'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders success correctly when the function resolves after 500ms', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn);
        return (
          <React.Fragment>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isResolved && 'success'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('success'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders error correctly when the function rejects after 500ms', async () => {
      const fn = jest.fn().mockReturnValue(new Promise((res, rej) => setTimeout(rej, 500)));
      const Component = () => {
        const testLoader = useLoads(fn);
        return (
          <React.Fragment>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isRejected && 'error'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('error'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });
  });

  describe('with deferred load', () => {
    it('renders loading indicator correctly (default delay)', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isPending && 'loading'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      fireEvent.click(getByText('load'));
      expect(fn).toBeCalledTimes(1);
      await waitForElement(() => getByText('loading'));
      expect(container).toMatchSnapshot();
    });

    it('renders loading indicator correctly (delay = 300)', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => res()));
      const Component = () => {
        const testLoader = useLoads(fn, { delay: 300, defer: true });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isPending && 'loading'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      fireEvent.click(getByText('load'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders timeout indicator correctly', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true, timeout: 400 });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isPending && 'loading'}
            {testLoader.isTimeout && 'timeout'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      fireEvent.click(getByText('load'));
      expect(fn).toBeCalledTimes(1);
      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('timeout'));
      expect(container).toMatchSnapshot();
    });

    it('renders success correctly when the function has been resolved', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => res()));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isResolved && 'success'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      expect(container).toMatchSnapshot();
      fireEvent.click(getByText('load'));
      expect(fn).toBeCalledTimes(1);
      await waitForElement(() => getByText('success'));
      expect(container).toMatchSnapshot();
    });

    it('renders success correctly when the function resolves after 500ms', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(res, 500)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isResolved && 'success'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      expect(container).toMatchSnapshot();
      fireEvent.click(getByText('load'));
      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('success'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders error correctly when the function rejects after 500ms', async () => {
      const fn = jest.fn().mockReturnValue(new Promise((res, rej) => setTimeout(rej, 500)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isIdle && 'idle'}
            {testLoader.isPending && 'loading'}
            {testLoader.isRejected && 'error'}
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      expect(fn).toBeCalledTimes(0);
      expect(container).toMatchSnapshot();
      fireEvent.click(getByText('load'));
      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('error'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });
  });
});

describe('context (cache)', () => {
  describe('with load on mount', () => {
    describe('on success', () => {
      it('initially renders & caches data correctly', async () => {
        const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(() => res('mockData'), 50)));
        const Component = () => {
          const testLoader = useLoads(fn, { context: 'foo-success', delay: 0 });
          return (
            <React.Fragment>
              {testLoader.isPending && <span>loading</span>}
              {testLoader.isResolved && <span>success</span>}
              <span>{testLoader.response}</span>
            </React.Fragment>
          );
        };

        const { container, getByText } = render(<Component />);

        await waitForElement(() => getByText('loading'));
        await waitForElement(() => getByText('success'));
        expect(fn).toBeCalledTimes(1);
        expect(container).toMatchSnapshot();
      });

      it('renders the cached data correctly on subsequent load', async () => {
        const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(() => res('mockData'), 50)));
        const Component = () => {
          const testLoader = useLoads(fn, { context: 'foo-success', delay: 0 });
          return (
            <React.Fragment>
              {testLoader.isPending && <span>loading</span>}
              {testLoader.isResolved && <span>success</span>}
              <span>{testLoader.response}</span>
            </React.Fragment>
          );
        };

        const { container, getByText } = render(<Component />);

        expect(fn).toBeCalledTimes(0);
        await waitForElement(() => getByText('success'));
        expect(container).toMatchSnapshot();
      });
    });

    describe('on error', () => {
      it('initially renders & caches data correctly', async () => {
        const fn = jest.fn().mockReturnValue(new Promise((res, rej) => setTimeout(() => rej('mockError'), 50)));
        const Component = () => {
          const testLoader = useLoads(fn, { context: 'foo-error', delay: 0 });
          return (
            <React.Fragment>
              {testLoader.isPending && <span>loading</span>}
              {testLoader.isRejected && <span>error</span>}
              <span>{testLoader.error}</span>
            </React.Fragment>
          );
        };

        const { container, getByText } = render(<Component />);

        await waitForElement(() => getByText('loading'));
        await waitForElement(() => getByText('error'));
        await waitForElement(() => getByText('mockError'));
        expect(fn).toBeCalledTimes(1);
        expect(container).toMatchSnapshot();
      });

      it('renders the cached data correctly on subsequent load', async () => {
        const fn = jest.fn().mockReturnValue(new Promise((res, rej) => setTimeout(() => rej('mockError'), 50)));
        const Component = () => {
          const testLoader = useLoads(fn, { context: 'foo-error', delay: 0 });
          return (
            <React.Fragment>
              {testLoader.isPending && <span>loading</span>}
              {testLoader.isRejected && <span>error</span>}
              <span>{testLoader.error}</span>
            </React.Fragment>
          );
        };

        const { container, getByText } = render(<Component />);

        expect(fn).toBeCalledTimes(0);
        await waitForElement(() => getByText('error'));
        await waitForElement(() => getByText('mockError'));
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('with deferred load', () => {
    it('initially renders & caches data correctly', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(() => res('mockData'), 50)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true, context: 'bar', delay: 0 });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isPending && <span>loading</span>}
            {testLoader.isResolved && <span>success</span>}
            <span>{testLoader.response}</span>
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      fireEvent.click(getByText('load'));
      await waitForElement(() => getByText('loading'));
      await waitForElement(() => getByText('success'));
      expect(fn).toBeCalledTimes(1);
      expect(container).toMatchSnapshot();
    });

    it('renders the cached data correctly on subsequent load', async () => {
      const fn = jest.fn().mockReturnValue(new Promise(res => setTimeout(() => res('mockData'), 50)));
      const Component = () => {
        const testLoader = useLoads(fn, { defer: true, context: 'bar', delay: 0 });
        return (
          <React.Fragment>
            <button onClick={testLoader.load}>load</button>
            {testLoader.isPending && <span>loading</span>}
            {testLoader.isResolved && <span>success</span>}
            <span>{testLoader.response}</span>
          </React.Fragment>
        );
      };

      const { container, getByText } = render(<Component />);

      fireEvent.click(getByText('load'));
      await waitForElement(() => getByText('success'));
      expect(container).toMatchSnapshot();
    });
  });
});
