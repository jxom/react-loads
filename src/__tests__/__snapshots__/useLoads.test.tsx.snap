// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context (cache) with deferred load initially renders & caches data correctly 1`] = `
<div>
  <button>
    load
  </button>
  <span>
    success
  </span>
  <span>
    mockData
  </span>
</div>
`;

exports[`context (cache) with deferred load renders the cached data correctly on subsequent load 1`] = `
<div>
  <button>
    load
  </button>
  <span>
    success
  </span>
  <span>
    mockData
  </span>
</div>
`;

exports[`context (cache) with load on mount on error initially renders & caches data correctly 1`] = `
<div>
  <span>
    error
  </span>
  <span>
    mockError
  </span>
</div>
`;

exports[`context (cache) with load on mount on error renders the cached data correctly on subsequent load 1`] = `
<div>
  <span>
    error
  </span>
  <span>
    mockError
  </span>
</div>
`;

exports[`context (cache) with load on mount on success initially renders & caches data correctly 1`] = `
<div>
  <span>
    success
  </span>
  <span>
    mockData
  </span>
</div>
`;

exports[`context (cache) with load on mount on success renders the cached data correctly on subsequent load 1`] = `
<div>
  <span>
    success
  </span>
  <span>
    mockData
  </span>
</div>
`;

exports[`states renders idle correctly the loading function has not been invoked 1`] = `
<div>
  idle
</div>
`;

exports[`states with deferred load renders error correctly when the function rejects after 500ms 1`] = `
<div>
  <button>
    load
  </button>
  idle
</div>
`;

exports[`states with deferred load renders error correctly when the function rejects after 500ms 2`] = `
<div>
  <button>
    load
  </button>
  error
</div>
`;

exports[`states with deferred load renders loading indicator correctly (default delay) 1`] = `
<div>
  <button>
    load
  </button>
  loading
</div>
`;

exports[`states with deferred load renders loading indicator correctly (delay = 0) 1`] = `
<div>
  <button>
    load
  </button>
</div>
`;

exports[`states with deferred load renders success correctly when the function has been resolved 1`] = `
<div>
  <button>
    load
  </button>
  idle
</div>
`;

exports[`states with deferred load renders success correctly when the function has been resolved 2`] = `
<div>
  <button>
    load
  </button>
  success
</div>
`;

exports[`states with deferred load renders success correctly when the function resolves after 500ms 1`] = `
<div>
  <button>
    load
  </button>
  idle
</div>
`;

exports[`states with deferred load renders success correctly when the function resolves after 500ms 2`] = `
<div>
  <button>
    load
  </button>
  success
</div>
`;

exports[`states with deferred load renders timeout indicator correctly 1`] = `
<div>
  <button>
    load
  </button>
  timeout
</div>
`;

exports[`states with load on initial render renders error correctly when the function rejects after 500ms 1`] = `
<div>
  error
</div>
`;

exports[`states with load on initial render renders loading indicator correctly (default delay) 1`] = `
<div>
  loading
</div>
`;

exports[`states with load on initial render renders success correctly when the function has been resolved 1`] = `
<div>
  success
</div>
`;

exports[`states with load on initial render renders success correctly when the function resolves after 500ms 1`] = `
<div>
  success
</div>
`;

exports[`states with load on initial render renders timeout indicator correctly 1`] = `
<div>
  timeout
</div>
`;
